### ISSUE TITLE
Accidental user actions such as changing file permissions should be reversible

### ISSUE THREAD

**Issue by 4jNsY6fCVqZv** [elementary/files#1121]
Unfortunately, I accidentally changed some permissions, but I can't remember which and how their original state was. Affects the Permissions tab in the file settings in the right-click menu.

**Comment by jeremypw** [comment]
There is code for undoing permissions already (with `<Ctrl>Z`) in the core library but it is not working when permissions are changed with the Properties window because that widget changes the file attributes directly, by-passing core code - which seems silly.  Never noticed that before - I have never had cause to undo a permission change.  It should not be too difficult to have the PropertiesWindow use core code for changing permissions.

Have to get UX team input on having an Undo button - there isn't one in the main window.

**Comment by 4jNsY6fCVqZv** [comment]
Thank you for thinking about it and finding a solution!

**Comment by 4jNsY6fCVqZv** [comment]
Thank you for taking care of the issue!

I'm also wondering: Are there other critical/sensitive actions in Files besides deleting files and giving access permissions that could benefit from an Undo function?

**Comment by jeremypw** [comment]
The undo manager only covers file operations at the moment - all of these (except permanent delete) should be covered.  Please let us know if you find any more that are not.  Other actions like create/delete bookmark are not considered critical and are not covered. Please raise an issue if there is something else you think should be undoable.

**Comment by 4jNsY6fCVqZv** [comment]
Do you think a solution makes sense in the context of https://github.com/elementary/files/pull/1132?

**Comment by jeremypw** [comment]
I do not think it is practical for Files to provide undo for external contractor apps.  I guess color tag undo might be feasible if there were a demand for it but it does not seem to be a priority.

**Comment by 4jNsY6fCVqZv** [comment]
I agree. What do you think of the first part of the comment?

**Comment by jeremypw** [comment]
When porting to Vala it is usually simpler to initially do a straight port unless the fix is very simple. I'll look into it.

**Comment by 4jNsY6fCVqZv** [comment]
Could you or someone else from the community imagine to work on a solution here?

**Comment by jeremypw** [comment]
This and many other issues it would be nice to fix but there are a lot of outstanding PRs awaiting review so there is little incentive to add to that queue.

**Comment by 4jNsY6fCVqZv** [comment]
@tintou Could you imagine to work on it?
